uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/implementation/ConnectionControlImpl.java

```diff
@@ -0,0 +1,137 @@
+import com.aws.greengrass.testing.mqtt.client.control.api.ConnectionControl;
+
+import java.util.Arrays;
+
+/**
+ * Implementation of ConnectionControl.
+ */
+public class ConnectionControlImpl implements ConnectionControl {
+
+    private final int connectionId;
+    private final Mqtt5ConnAck connAck;
+    private final ConnectionEvents connectionEvents;
+    private final AgentControlImpl agent;
+    private int timeout;
+
+    /**
+     * Creates instanse of ConnectionControlImpl.
+     * @param connectReply response to connect request from agent
+     * @param agent backreference to agent
+     */
+    ConnectionControlImpl(MqttConnectReply connectReply, ConnectionEvents connectionEvents, AgentControlImpl agent) {
+        super();
+        this.connectionId = connectReply.getConnectionId().getConnectionId();
+        this.connAck = connectReply.getConnAck();
+        this.connectionEvents = connectionEvents;
+        this.agent = agent;
+        this.timeout = agent.getTimeout();
+    }
+
+    @Override
+    public int getConnectionId() {
+        return connectionId;
+    }
+
+    @Override
+    public Mqtt5ConnAck getConnAck() {
+        return connAck;
+    }
+
+    @Override
+    public int getTimeout() {
+        return timeout;
+    }
+
+    @Override
+    public void setTimeout(int timeout) {
+        this.timeout = timeout;
+    }
+
+    /**
+     * Called when MQTT message has been received.
+     *
+     * @param message the received MQTT message
+     */
+    public void onMessageReceived(Mqtt5Message message) {
+        if (connectionEvents != null) {
+            connectionEvents.onMessageReceived(this, message);
+        }
+    }
+
+    /**
+     * Called when MQTT connection has been disconnected.
+     *
+     * @param disconnect optional infomation from DISCONNECT packet
+     * @param error optional OS-dependent error string
+     */
+    public void onMqttDisconnect(Mqtt5Disconnect disconnect, String error) {
+        if (connectionEvents != null) {
+            connectionEvents.onMqttDisconnect(this, disconnect, error);
+        }
+    }
+
+    @Override
+    public void closeMqttConnection(int reason) {
+        MqttCloseRequest closeRequest = MqttCloseRequest.newBuilder()
+                    .setConnectionId(MqttConnectionId.newBuilder().setConnectionId(connectionId).build())
+                    .setTimeout(timeout)
+                    .setReason(reason)
+                    .build();
+        agent.closeMqttConnection(closeRequest);
+    }
+
+    @Override
+    public MqttSubscribeReply subscribeMqtt(Integer subscriptionId, Mqtt5Subscription... subscriptions) {
+        MqttSubscribeRequest.Builder builder = MqttSubscribeRequest.newBuilder()
+                .setConnectionId(MqttConnectionId.newBuilder().setConnectionId(connectionId).build())
+                .setTimeout(timeout)
+                .addAllSubscriptions(Arrays.asList(subscriptions));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;subscriptions&#39; (index: 1 | type: Mqtt5Subscription[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;subscriptionId&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).

Reply:
resolved in 42c58d2a9e4a8e05096433ce6040f96b564cf04c